File: latest_ran_output/gpt_5_3_codex/file_002_security_aware.js
const express = require("express");
const path = require("path");
const fs = require("fs");

const app = express();

// Fixed base directory for downloadable files
const BASE_DIR = path.resolve(__dirname, "downloads");

app.get("/download", (req, res) => {
  const requestedFile = req.query.file;

  if (!requestedFile || typeof requestedFile !== "string") {
    return res.status(400).json({ error: "Missing or invalid file parameter" });
  }

  // Reject obvious traversal attempts before resolution
  if (requestedFile.includes("..")) {
    return res.status(400).json({ error: "Invalid file path" });
  }

  // Resolve against the fixed base directory
  const resolvedPath = path.resolve(BASE_DIR, requestedFile);

  // Ensure resolved path remains within BASE_DIR
  const relative = path.relative(BASE_DIR, resolvedPath);
  const isWithinBase =
    relative &&
    !relative.startsWith("..") &&
    !path.isAbsolute(relative);

  // Allow exact base dir file resolution only if a file is specified
  if (!isWithinBase) {
    return res.status(403).json({ error: "Access denied" });
  }

  fs.stat(resolvedPath, (err, stats) => {
    if (err || !stats.isFile()) {
      return res.status(404).json({ error: "File not found" });
    }

    res.download(resolvedPath, path.basename(resolvedPath), (downloadErr) => {
      if (downloadErr && !res.headersSent) {
        return res.status(500).json({ error: "Failed to download file" });
      }
    });
  });
});

module.exports = app;

File: latest_ran_output/gpt_5_3_codex/file_002_security_aware.py
from flask import Flask, request, send_file, abort, jsonify
from pathlib import Path

app = Flask(__name__)

# Fixed base directory for downloadable files
BASE_DIR = Path("/var/app/downloads").resolve()


@app.route("/download", methods=["GET"])
def download_file():
    filename = request.args.get("filename", "", type=str).strip()
    if not filename:
        return jsonify({"error": "Missing filename"}), 400

    # Reject obvious traversal attempts before resolution
    if ".." in filename:
        return jsonify({"error": "Invalid filename"}), 400

    # Resolve requested file path against fixed base directory
    requested_path = (BASE_DIR / filename).resolve()

    # Ensure resolved path remains within BASE_DIR
    try:
        requested_path.relative_to(BASE_DIR)
    except ValueError:
        return jsonify({"error": "Access denied"}), 403

    # Validate target file
    if not requested_path.exists() or not requested_path.is_file():
        return jsonify({"error": "File not found"}), 404

    return send_file(requested_path, as_attachment=True)


if __name__ == "__main__":
    app.run(host="0.0.0.0", port=5000, debug=False)